json.go

```diff
@@ -16,66 +16,74 @@
 package gomatrixserverlib
 
 import (
-	"bytes"
 	"encoding/binary"
-	"encoding/json"
 	"sort"
 	"unicode/utf8"
+
+	"github.com/pkg/errors"
+	"github.com/tidwall/gjson"
 )
 
 // CanonicalJSON re-encodes the JSON in a canonical encoding. The encoding is
 // the shortest possible encoding using integer values with sorted object keys.
 // https://matrix.org/docs/spec/server_server/unstable.html#canonical-json
 func CanonicalJSON(input []byte) ([]byte, error) {
-	sorted, err := SortJSON(input, make([]byte, 0, len(input)))
-	if err != nil {
-		return nil, err
+	if !gjson.Valid(string(input)) {
+		return nil, errors.Errorf("invalid json")
 	}
-	return CompactJSON(sorted, make([]byte, 0, len(sorted))), nil
+
+	input = CompactJSON(input, make([]byte, 0, len(input)))
+	return SortJSON(input, make([]byte, 0, len(input))), nil
+}
+
+// CanonicalJSONAssumeValid is the same as CanonicalJSON, but assumes the
+// input is valid JSON
+func CanonicalJSONAssumeValid(input []byte) []byte {
+	input = CompactJSON(input, make([]byte, 0, len(input)))
```

Comment:
I'm wondering why we used to Sort and then Compact, and now we Compact and then Sort. It looks like SortJSON removes extraneous whitespace, so perhaps it's more efficient to Compact after Sorting? (In fact, should we combine the two by doing string compaction when the input is a string in `sortJSONValue`?)

Reply:
To be honest, it was mainly that it made it easier debugging for various spurious reasons. I don't think it matters one way or the other.

> In fact, should we combine the two by doing string compaction when the input is a string in sortJSONValue

Quite possibly, canonicalisation takes about 10%